Quadtree bounding box 구현

rnfxl92 opened the pull request:
## 구현내용
### [feat], [clustering], [test]
* BoundingBox 네개로 쪼개기 구현했습니다.
* contains(coordinate: Coordinate) 구현했습니다.
* intersectsBoxBounds(with box: BoundingBox) 구현했습니다.
* 테스트 했습니다.

## 논의사항
Coordinate, Cluster, BoundingBox 을 각 알고리즘이 같이 사용하게끔 프로토콜 협의 필요합니다.

rnfxl92 committed:
[feat] QuadTree에 사용할 BoundingBox 구현

 
- 각 함수 구현
- Cluster, Coordinate 구조체는 현재 mockup 구조체임

rnfxl92 committed:
[feat] BoundingBox Test 추가

beggu84 reviewed (commented):
프로젝트에 QuadTree 폴더가 빠져있는 것 같아요
![image](https://user-images.githubusercontent.com/19990608/100087286-dc887c80-2e91-11eb-8789-e7930445be7a.png)

rnfxl92 committed:
[fix] project File 추가

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/QuadTree/BoundingBox.swift:
```diff
@@ -0,0 +1,58 @@
+        let midX: Double = (bottomLeft.x + topRight.x) / 2.0
+        let midY: Double = (bottomLeft.y + topRight.y) / 2.0
+        return Coordinate(x: midX, y: midY)
+    }
```
혹시 computed property 쓰면 mid 호출할때마다 계속 Coordinate를 생성하나요 ?

eunjeongS2 replied:
리뷰 반영 했습니다! 확인해주세요 🥺

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/QuadTree/BoundingBox.swift:
```diff
@@ -0,0 +1,58 @@
+    
+    func intersectsBoxBounds(with box: BoundingBox) -> Bool {
+        return (bottomLeft <= box.topRight &&  box.bottomLeft <= topRight)
+    }
```
뭔가 bool을 반환하는 함수명같지않아요

eunjeongS2 replied:
isOverlapped로 변경했습니다. 🥺

eunjeongS2 replied:
근데 앞에 is를 붙이는건지 안붙이는건지 쫌 헷갈리네요 🤔

A-by-alimelon replied:
box의 topLeft 가 bottomRight 랑 겹치는 경우는 생각 안해도 되는 건가요 ??

eunjeongS2 replied:
@A-by-alimelon @beggu84 기존코드에 추가 테스트코드를 넣었는데 잘 작동했습니다 !

beggu84 replied:
매개변수 명을 `other` 로 하면, 내용도 `bottomLeft <= other.topRight &&  other.bottomLeft <= topRight` 가 되서 가독성이 높아질 것 같습니다.
하지만, 그렇게 하면 `contains` 메서드의 매개변수 명은 어떻게 해야하나~ 까지 고민하게 될 수도 있으므로, 무시하셔도 좋습니다^^

donggeonoh commented on Project17-C-Map/InteractiveClusteringMap/QuadTree/BoundingBox.swift:
```diff
@@ -0,0 +1,58 @@
+        return (containsX && containsY)
+    }
+    
+    func intersectsBoxBounds(with box: BoundingBox) -> Bool {
+        return (bottomLeft <= box.topRight &&  box.bottomLeft <= topRight)
+    }
```
메서드 네임 앞에 is 가 붙으면 더 좋을것같아요!

eunjeongS2 replied:
isOverlapped로 변경했습니다~~ 어떠신가용

donggeonoh replied:
네 저는 좋아보여요 ㅋㅋㅋ

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/QuadTree/BoundingBox.swift:
```diff
@@ -0,0 +1,58 @@
+            BoundingBox(topRight: Coordinate(x: topRight.x, y: mid.y),
+                        bottomLeft: Coordinate(x: mid.x, y: bottomLeft.y))
+        ]
+    }
```
computed property

eunjeongS2 replied:
이것도 바꿨습니다! 확인해주세용

A-by-alimelon reviewed (dismissed):
LGTM

rnfxl92 commented:
@beggu84  프로젝트 파일 푸시했습니다. 다시 확인 부탁드립니다!!

rnfxl92 committed:
[refactor] 리뷰반영

 
- computed property -> lazy var로 변경
- 메서드 이름변경

rnfxl92 committed:
[test] test case 추가

rnfxl92 committed:
[refactor] Coordinate Center

 
- Center 구하는 함수 Box -> Coordinate로 이동
- splitedQuadBoundingBox lazy var에서 func로 변경
lazy var일경우 변경될 수 있기 때문에 상수화를 위해 func으로 변경

beggu84 commented:
테스트케이스명이 한글이라 가독성이 좋네요.
하지만, 본인의 발전을 위해서 영어로 쓰는 습관을 들이면 더 좋을 것 같습니다.
(알고보니 네이티브시고..)

rnfxl92 committed:
[test] test case 추가

rnfxl92 commented:
@beggu84 다음 테스트부터는 영어로 쓰겠습니다. other를 쓰면 가독성이 더 좋아질 것같다는 리뷰도 반영했습니다. 또한 가독성을 위해 self도 붙이는게 좋겠다는 팀원의 리뷰가 있어서 self도 붙였는데 self를 붙여도 괜찮을까요?

이 늦은 시간에도 리뷰해주셔서 감사합니다 ~~👍
